### PR TITLE
echo -> printf

### PR DESCRIPTION
**Pull request by LOAS2**
echo can have unspecified behavior which is very undesirable.
To stay on the safe side, it's good practice to always use printf.
echo "$var" can sting you whenever you do not have strict guarantees about the contents of var (for example that it contains a nonnegative integer).

READ - http://www.etalabs.net/sh_tricks.html

### CODE REVIEW

File: contrib/windows/get_toolchain.sh [hunk 1]
```diff
@@ -22,11 +22,11 @@ case $bits in
     exc=seh
     ;;
   *)
-    echo 'error: run script either as `./get_toolchain.sh 32` or `./get_toolchain.sh 64`' >&2
+    printf "error: run script either as `./get_toolchain.sh 32` or `./get_toolchain.sh 64`\n" >&2
```
Comment (tkelman) [inline contrib/windows/get_toolchain.sh]: will changing the quotes here make the backticks get run as code?

File: contrib/windows/msys_build.sh [hunk 2]
```diff
@@ -31,49 +31,49 @@ checksum_download() {
...
+  printf "LIBBLASNAME = libopenblas64_\n" >> Make.user
 else
   bits=32
   archsuffix=86
   exc=sjlj
-  echo "override MARCH = pentium4" >> Make.user
-  echo 'LIBBLAS = -L$(JULIAHOME)/usr/bin -lopenblas' >> Make.user
-  echo 'LIBBLASNAME = libopenblas' >> Make.user
+  printf "override MARCH = pentium4\n" >> Make.user
+  printf "LIBBLAS = -L$(JULIAHOME)/usr/bin -lopenblas\n" >> Make.user
```
Comment (tkelman) [inline contrib/windows/msys_build.sh]: the single quotes are important here, JULIAHOME is meant as the makefile variable and should not be expanded yet

File: contrib/windows/msys_build.sh [hunk 3]
```diff
@@ -31,49 +31,49 @@ checksum_download() {
...
-      echo "XC_HOST = $XC_HOST" >> Make.user
+      printf "XC_HOST = $XC_HOST\n" >> Make.user
     fi
     CROSS_COMPILE="$XC_HOST-"
     # Set BUILD_MACHINE and HOSTCC in case we don't have Cygwin gcc installed
-    echo "override BUILD_MACHINE = $ARCH-pc-cygwin" >> Make.user
+    printf "override BUILD_MACHINE = $ARCH-pc-cygwin\n" >> Make.user
     if [ -z "`which gcc 2>/dev/null`" ]; then
-      echo 'override HOSTCC = $(CROSS_COMPILE)gcc' >> Make.user
+      printf "override HOSTCC = $(CROSS_COMPILE)gcc\n" >> Make.user
```
Comment (tkelman) [inline contrib/windows/msys_build.sh]: this would have to change from makefile variable to shell variable

File: contrib/windows/msys_build.sh [hunk 4]
```diff
@@ -125,35 +125,35 @@ if [ -z "$USEMSVC" ]; then
...
+  printf "override XC_HOST = \n" >> Make.user
   export CC="$PWD/deps/srccache/libuv/compile cl -nologo -MD -Z7"
   export AR="$PWD/deps/srccache/libuv/ar-lib lib"
   export LD="$PWD/linkld link"
-  echo "override CC = $CC" >> Make.user
-  echo 'override CXX = $(CC) -EHsc' >> Make.user
-  echo "override AR = $AR" >> Make.user
-  echo "override LD = $LD -DEBUG" >> Make.user
+  printf "override CC = $CC\n" >> Make.user
+  printf "override CXX = $(CC) -EHsc\n" >> Make.user
```
Comment (tkelman) [inline contrib/windows/msys_build.sh]: also makefile variable, single quotes mattered

File: contrib/windows/msys_build.sh [hunk 5]
```diff
@@ -163,49 +163,49 @@ fi
...
 
 for lib in SUITESPARSE ARPACK BLAS LAPACK FFTW \
     GMP MPFR PCRE LIBUNWIND OPENSPECFUN; do
-  echo "USE_SYSTEM_$lib = 1" >> Make.user
+  printf "USE_SYSTEM_$lib = 1\n" >> Make.user
 done
-echo 'override LIBLAPACK = $(LIBBLAS)' >> Make.user
-echo 'override LIBLAPACKNAME = $(LIBBLASNAME)' >> Make.user
+printf "override LIBLAPACK = $(LIBBLAS)\n" >> Make.user
+printf "override LIBLAPACKNAME = $(LIBBLASNAME)\n" >> Make.user
```
Comment (tkelman) [inline contrib/windows/msys_build.sh]: and here

**Comment by StefanKarpinski** [comment]
I have this suspicion that this might be one of those cases where `echo` is in theory less portable, but in practice more portable because all the different versions behave similarly enough. But that's just a hunch. Is this fixing an actual problem on some system that you're using?

**Comment by LOAS2** [comment]
This isn't fixing any problems for me.
But printf is always a better choice like shown - http://www.etalabs.net/sh_tricks.html
You’ll run into trouble someday when you happen to read a “-n” or “-e” or “-neEenenEene” into var and suddenly your script breaks.

Better keep good practice - always use printf and be on the safe side.
I mean - why not? I only see advantages.

File: contrib/windows/msys_build.sh [hunk 6]
```diff
@@ -31,49 +31,49 @@ checksum_download() {
...
   bits=64
   archsuffix=64
   exc=seh
-  echo "override MARCH = x86-64" >> Make.user
-  echo 'USE_BLAS64 = 1' >> Make.user
-  echo 'LIBBLAS = -L$(JULIAHOME)/usr/bin -lopenblas64_' >> Make.user
-  echo 'LIBBLASNAME = libopenblas64_' >> Make.user
+  printf "override MARCH = x86-64\n" >> Make.user
+  printf "USE_BLAS64 = 1\n" >> Make.user
+  printf "LIBBLAS = -L$(JULIAHOME)/usr/bin -lopenblas64_\n" >> Make.user
```
Comment (tkelman) [inline contrib/windows/msys_build.sh]: should stay single quotes

**Comment by StefanKarpinski** [comment]
> I mean - why not? I only see advantages.

The obvious problem would be if there are systems on which `printf` doesn't exist, or on which `printf` has non-standard behavior. I suppose we can try this and see how it goes.

**Comment by LOAS2** [comment]
printf is defined by POSIX. All systems should have it.
echo is actually the problem. According to POSIX, echo has unspecified behavior in some cases.

**Comment by ViralBShah** [comment]
How come not a single build is passing CI?

**Comment by tkelman** [comment]
needs a rebase, there's a conflict with some changes I made to one of these scripts

**Comment by TotalVerb** [comment]
Just blanket-replacing `echo` with `printf` does nothing to avoid its problems.

Consider for example this line:

```sh
printf "$providers\n"
```

Now suppose the value `providers` was `%d`... what happens?

```
0
```

The advice of using `printf` instead of `echo` is good, but this is the incorrect way to approach the change. `printf` is not a magical potion that is a drop-in replacement for echo. The _correct_ way is to do

```sh
printf '%s\n' "$providers"
```

which is indeed safe, no matter the contents of the `providers` variable. Unfortunately, this is not a change that can be made with find and replace.

**Comment by yuyichao** [comment]
It seems that many of the changes are unnecessary (especially ones that just print a literal string)

Some of them shouldn't use echo/printf at all e.g. https://github.com/JuliaLang/julia/pull/18158/files#diff-6aa1e3cf76ed6caa6bb4f23dac0d2bb4L71

Some should not have `\n` in additional to adding `%` handling. e.g. https://github.com/JuliaLang/julia/pull/18158/files#diff-2d001c34a7f446af67f69d0348519905L10

Also almost none of the changes actually avoid any undesired behavior since almost all (apart from 3-4) uses are guaranteed to not start with `-` and instead of the "undefined" handling of `\` in `echo`, `printf` always gives the undesired behavior when `\` appears in the string.

**Comment by yuyichao** [comment]
I think the right way to fix this PR is to

1. For cases where there isn't any problem (generally large number of literal strings) e.g. [this](https://github.com/JuliaLang/julia/pull/18158/files#diff-36e175b02ff75b8431a633c4bb139219) or [this](https://github.com/JuliaLang/julia/pull/18158/files#diff-026b4999efcfd32f3f63bae007b94e31), leave it as what it was before.

2. Fix the cases where the additional wrapping isn't necessary, e.g. [this](https://github.com/JuliaLang/julia/pull/18158/files#diff-173592af8fce0214f7368aae0d5139ebL12) in additional to the case mentioned above in the previous comment

3. For a few of the long script that could trigger undesired behavior of `echo` (typically from path names), define a wrapper function that prints a literal string and use that instead.

4. For script where `echo`ing of potentially problematic interpolated string is only used a few time, either using the approach in 3 or use `printf` formatting directly should work.

**Comment by LOAS2** [comment]
I was a noob who read about the echo problems and saw the printf solution.
I thank TotalVerb and yuyichao for showing a simple search-replace isn't the right way to fix the problems.
So I will close this PR and open an issue instead so experts can work on this, and fix in the right way.
I find it good that at least the issue was recognized and people agree echo can be dangerous in some places.
Thanks all.

**Comment by StefanKarpinski** [comment]
Thanks for bringing the issue to our attention, @LOAS2.

**Pull request closed without merging**